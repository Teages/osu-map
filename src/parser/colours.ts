export function parseColours(lines: string[]): Colours {
  const colours: Colours = {}

  lines.forEach((line) => {
    const [name, value] = line.split(':').map(v => v.trim())
    if (name === 'SliderTrackOverride') {
      colours.SliderTrackOverride = parseColor(value)
    }
    else if (name === 'SliderBorder') {
      colours.SliderBorder = parseColor(value)
    }
    else if (name.startsWith('Combo')) {
      const index = Number(name.slice('Combo'.length))
      colours[index] = parseColor(value)
    }
  })

  return colours
}

function parseColor(value: string): Color {
  const [red, green, blue] = value.split(',').map(v => v.trim()).map(v => Number(v))
  return {
    red,
    green,
    blue,
    hex: `#${red.toString(16).padStart(2, '0')}${green.toString(16).padStart(2, '0')}${blue.toString(16).padStart(2, '0')}`,
  }
}

export interface Colours {
  /**
   * Additive combo colours
   */
  [key: number]: Color

  /**
   * Additive slider track colour
   */
  SliderTrackOverride?: Color

  /**
   * Slider border colour
   */
  SliderBorder?: Color
}
export interface Color {
  red: number
  green: number
  blue: number
  hex: string
}
