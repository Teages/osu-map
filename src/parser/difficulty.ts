import { useKVPairParser } from '../utils'

export function parseDifficulty(input: string[]): Difficulty {
  const typeMap = {
    HPDrainRate: 'number',
    CircleSize: 'number',
    OverallDifficulty: 'number',
    ApproachRate: 'number',
    SliderMultiplier: 'number',
    SliderTickRate: 'number',
  } as const

  const defaultValues = {
    HPDrainRate: 0,
    CircleSize: 0,
    OverallDifficulty: 0,
    ApproachRate: 0,
    SliderMultiplier: 1,
    SliderTickRate: 1,
  }

  const difficultyParser = useKVPairParser<Difficulty>(typeMap, defaultValues)

  return difficultyParser(input)
}

export interface Difficulty {
  /**
   * HP setting (0–10)
   */
  HPDrainRate: number

  /**
   * CS setting (0–10)
   */
  CircleSize: number

  /**
   * OD setting (0–10)
   */
  OverallDifficulty: number

  /**
   * AR setting (0–10)
   */
  ApproachRate: number

  /**
   * Base slider velocity in hundreds of [osu! pixels](https://osu.ppy.sh/wiki/Client/Beatmap_editor/osu!_pixel) per beat
   */
  SliderMultiplier: number

  /**
   * Amount of slider ticks per beat
   */
  SliderTickRate: number
}
