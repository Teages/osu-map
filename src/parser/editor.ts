import { useKVPairParser } from '../utils'

export function parseEditor(input: string[]): Editor {
  const typeMap = {
    Bookmarks: 'string',
    DistanceSpacing: 'number',
    BeatDivisor: 'number',
    GridSize: 'number',
    TimelineZoom: 'number',
  } as const

  const defaultValues = {
    Bookmarks: '',
  }

  const editorParser = useKVPairParser<Omit<Editor, 'Bookmarks'> & { Bookmarks: string }>(typeMap, defaultValues)

  const res = editorParser(input)

  return {
    ...res,
    Bookmarks: res.Bookmarks.split(',').map(v => v.trim()).map(Number),
  }
}

export interface Editor {
  /**
   * Time in milliseconds of [bookmarks](https://osu.ppy.sh/wiki/Client/Beatmap_editor/Compose#song-timeline)
   */
  Bookmarks: number[]

  /**
   * [Distance snap](https://osu.ppy.sh/wiki/Client/Beatmap_editor/Distance_snap) multiplier
   */
  DistanceSpacing?: number

  /**
   * [Beat snap divisor](https://osu.ppy.sh/wiki/Client/Beatmap_editor/Beat_snap_divisor)
   */
  BeatDivisor?: number

  /**
   * [Grid size](https://osu.ppy.sh/wiki/Beatmapping/Grid_snapping)
   */
  GridSize?: number

  /**
   * Scale factor for the [object timeline](https://osu.ppy.sh/wiki/Client/Beatmap_editor/Compose#hit-object-timeline)
   */
  TimelineZoom?: number
}
