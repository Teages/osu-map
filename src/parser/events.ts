export function parseEvents(input: string[]): Events {
  const rawEvents = input.map((line) => {
    const [eventType, startTime, ...eventParams] = line.split(',')
    const parsedEventType = Number(eventType)
    return {
      eventType: Number.isNaN(parsedEventType) ? eventType : parsedEventType,
      startTime: Number(startTime.trim()),
      eventParams: eventParams.map(v => v.trim()),
    }
  })

  const rawBackground = rawEvents.find(e => e.eventType === 0)
  const rawVideo = rawEvents.find(e => e.eventType === 'Video' || e.eventType === 1)
  const breaks = rawEvents.filter(e => e.eventType === 'Break' || e.eventType === 2)
    .map(e => ({
      startTime: e.startTime,
      endTime: Number(e.eventParams[0]),
    }))

  return {
    background: rawBackground
      ? {
          filename: cleanFilename(rawBackground.eventParams[0]),
          xOffset: Number(rawBackground.eventParams[1]) || 0,
          yOffset: Number(rawBackground.eventParams[2]) || 0,
        }
      : undefined,
    video: rawVideo
      ? {
          startTime: rawVideo.startTime,
          filename: cleanFilename(rawVideo.eventParams[0]),
          xOffset: Number(rawVideo.eventParams[1]) || 0,
          yOffset: Number(rawVideo.eventParams[2]) || 0,
        }
      : undefined,
    breaks,
    raw: rawEvents,
  }
}

// Double quotes are usually included surrounding the filename, but they are not required.
function cleanFilename(filename: string) {
  if (filename.startsWith('"') && filename.endsWith('"')) {
    return filename.slice(1, -1)
  }
  return filename
}

export interface Events {
  /**
   * Background
   */
  background?: {
    /**
     * Location of the background image relative to the beatmap directory. Double quotes are usually included surrounding the filename, but they are not required.
     *
     * The library will automatically remove the double quotes.
     */
    filename: string

    /**
     * Horizontal offset of the background image. Using [osu! pixels](https://osu.ppy.sh/wiki/Client/Beatmap_editor/osu!_pixel) as the unit.
     */
    xOffset: number

    /**
     * Vertical offset of the background image. Using [osu! pixels](https://osu.ppy.sh/wiki/Client/Beatmap_editor/osu!_pixel) as the unit.
     */
    yOffset: number
  }

  /**
   * Video events
   */
  video?: {
    /**
     * Time when the video starts, in milliseconds from the beginning of the beatmap's audio.
     */
    startTime: number

    /**
     * Location of the background video relative to the beatmap directory. Double quotes are usually included surrounding the filename, but they are not required.
     *
     * The library will automatically remove the double quotes.
     */
    filename: string

    /**
     * Horizontal offset of the background image. Using [osu! pixels](https://osu.ppy.sh/wiki/Client/Beatmap_editor/osu!_pixel) as the unit.
     */
    xOffset: number

    /**
     * Vertical offset of the background image. Using [osu! pixels](https://osu.ppy.sh/wiki/Client/Beatmap_editor/osu!_pixel) as the unit.
     */
    yOffset: number
  }

  /**
   * Break events
   */
  breaks: Array<{
    /**
     * Time when the break starts, in milliseconds from the beginning of the beatmap's audio.
     */
    startTime: number

    /**
     * Time when the break ends, in milliseconds from the beginning of the beatmap's audio.
     */
    endTime: number
  }>

  /**
   * Raw event data
   */
  raw: RawEvent[]
}
export interface RawEvent {
  /**
   * Type of the event. Some events may be referred to by either a name or a number.
   */
  eventType: number | string

  /**
   * Start time of the event, in milliseconds from the beginning of the beatmap's audio. For events that do not use a start time, the default is `0`.
   */
  startTime: number

  /**
   * Extra parameters specific to the event's type.
   */
  eventParams: Array<number | string>
}
