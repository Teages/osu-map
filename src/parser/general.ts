import { useKVPairParser } from '../utils'

export function parseGeneral(input: string[]): General {
  const typeMap: Record<keyof General, 'string' | 'number'> = {
    AudioFilename: 'string',
    AudioLeadIn: 'number',
    AudioHash: 'string',
    PreviewTime: 'number',
    Countdown: 'number',
    SampleSet: 'string',
    StackLeniency: 'number',
    Mode: 'number',
    LetterboxInBreaks: 'number',
    StoryFireInFront: 'number',
    UseSkinSprites: 'number',
    AlwaysShowPlayfield: 'number',
    OverlayPosition: 'string',
    SkinPreference: 'string',
    EpilepsyWarning: 'number',
    CountdownOffset: 'number',
    SpecialStyle: 'number',
    WidescreenStoryboard: 'number',
    SamplesMatchPlaybackRate: 'number',
  } as const

  const defaultValues = {
    AudioFilename: '',
    AudioLeadIn: 0,
    PreviewTime: -1,
    Countdown: 1,
    SampleSet: 'Normal',
    StackLeniency: 0.7,
    Mode: 0,
    LetterboxInBreaks: 0,
    StoryFireInFront: 1,
    UseSkinSprites: 0,
    AlwaysShowPlayfield: 0,
    OverlayPosition: 'NoChange',
    EpilepsyWarning: 0,
    CountdownOffset: 0,
    SpecialStyle: 0,
    WidescreenStoryboard: 0,
    SamplesMatchPlaybackRate: 0,
  }

  const generalParser = useKVPairParser<General>(typeMap, defaultValues)

  return generalParser(input)
}

export interface General {
  /**
   * Location of the audio file relative to the current folder
   */
  AudioFilename: string

  /**
   * Milliseconds of silence before the audio starts playing
   *
   * @default 0
   */
  AudioLeadIn: number

  /**
   * @deprecated
   */
  AudioHash?: string

  /**
   * Time in milliseconds when the audio preview should start
   *
   * @default -1
   */
  PreviewTime: number

  /**
   * Speed of the countdown before the first hit object
   *
   * `0` = no countdown, `1` = normal, `2` = half, `3` = double
   *
   * @default 1
   */
  Countdown: 0 | 1 | 2 | 3

  /**
   * Sample set that will be used if timing points do not override it
   *
   * `Normal`, `Soft`, `Drum`
   *
   * @default Normal
   */
  SampleSet: 'Normal' | 'Soft' | 'Drum'

  /**
   * Multiplier for the threshold in time where hit objects placed close together stack (0â€“1)
   *
   * @default 0.7
   */
  StackLeniency: number

  /**
   * Game mode
   *
   * `0` = osu!, `1` = osu!taiko, `2` = osu!catch, `3` = osu!mania
   *
   * @default 0
   */
  Mode: OsuMapMode

  /**
   * Whether or not breaks have a letterboxing effect
   *
   * @default 0
   */
  LetterboxInBreaks: number

  /**
   * @deprecated
   *
   * @default 1
   */
  StoryFireInFront: number

  /**
   * Whether or not the storyboard can use the user's skin images
   *
   * @default 0
   */
  UseSkinSprites: number

  /**
   * @deprecated
   *
   * @default 0
   */
  AlwaysShowPlayfield: number

  /**
   * Draw order of hit circle overlays compared to hit numbers
   *
   * `NoChange` = use skin setting, `Below` = draw overlays under numbers, `Above` = draw overlays on top of numbers
   *
   * @default NoChange
   */
  OverlayPosition: 'NoChange' | 'Below' | 'Above'

  /**
   * Preferred skin to use during gameplay
   */
  SkinPreference?: string

  /**
   * Whether or not a warning about flashing colours should be shown at the beginning of the map
   *
   * @default 0
   */
  EpilepsyWarning: number

  /**
   * Time in beats that the countdown starts before the first hit object
   *
   * @default 0
   */
  CountdownOffset: number

  /**
   * Whether or not the "N+1" style key layout is used for osu!mania
   *
   * @default 0
   */
  SpecialStyle: number

  /**
   * Whether or not the storyboard allows widescreen viewing
   *
   * @default 0
   */
  WidescreenStoryboard: number

  /**
   * Whether or not sound samples will change rate when playing with speed-changing mods
   *
   * @default 0
   */
  SamplesMatchPlaybackRate: number
}

export enum OsuMapMode {
  Standard = 0,
  Taiko = 1,
  Catch = 2,
  Mania = 3,
}
