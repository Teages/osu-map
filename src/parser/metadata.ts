import { useKVPairParser } from '../utils'

export function parseMetadata(input: string[]): Metadata {
  const typeMap = {
    Title: 'string',
    TitleUnicode: 'string',
    Artist: 'string',
    ArtistUnicode: 'string',
    Creator: 'string',
    Version: 'string',
    Source: 'string',
    Tags: 'string',
    BeatmapID: 'number',
    BeatmapSetID: 'number',
  } as const

  const generalParser = useKVPairParser<Omit<Metadata, 'Tags'> & { Tags: string }>(typeMap, {})

  const res = generalParser(input)
  return {
    ...res,
    Tags: res.Tags.split(' '),
  }
}

export interface Metadata {
  /**
   * Romanised song title
   */
  Title?: string

  /**
   * Song title
   */
  TitleUnicode?: string

  /**
   * Romanised song artist
   */
  Artist?: string

  /**
   * Song artist
   */
  ArtistUnicode?: string

  /**
   * Beatmap creator
   */
  Creator?: string

  /**
   * Difficulty name
   */
  Version?: string

  /**
   * Original media the song was produced for
   */
  Source?: string

  /**
   * Search terms
   */
  Tags?: string[]

  /**
   * Difficulty ID
   */
  BeatmapID?: number

  /**
   * Beatmap ID
   */
  BeatmapSetID?: number
}
