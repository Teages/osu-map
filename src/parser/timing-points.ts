export function parseTimingPoints(input: string[]): TimingPoint[] {
  const timingPoints: TimingPoint[] = []

  input.forEach((line) => {
    const [
      time,
      beatLength,
      meter,
      sampleSet,
      sampleIndex,
      volume,
      uninherited,
      effects,
    ] = line.split(',').map(v => v.trim()).map(v => Number(v))
    const item: TimingPoint = {
      time,
      meter,
      sampleSet,
      sampleIndex,
      volume,
      effect: {
        kiai: (effects & 1) === 1,
        omitFirstBarLine: (effects & 8) === 8,
      },
      ...(
        uninherited === 1
          ? {
              type: 'uninherited',
              BPM: Math.round(60000 / beatLength * 1e6) / 1e6,
            }
          : {
              type: 'inherited',
              sliderMultiplier: Math.round(-100 / beatLength * 1e6) / 1e6,
            }
      ),
    }

    timingPoints.push(item as TimingPoint)
  })

  return timingPoints
}

export type TimingPoint = {
  /**
   * Start time of the timing section, in milliseconds from the beginning of the beatmap's audio. The end of the timing section is the next timing point's time (or never, if this is the last timing point).
   */
  time: number

  /**
   * Amount of beats in a measure. Inherited timing points ignore this property.
   */
  meter: number

  /**
   * Default sample set for hit objects.
   *
   * `0` = beatmap default, `1` = normal, `2` = soft, `3` = drum
   */
  sampleSet: SampleSet

  /**
   * Custom sample index for hit objects. `0` indicates osu!'s default hitsounds.
   */
  sampleIndex: number

  /**
   * Volume percentage for hit objects.
   */
  volume: number

  effect: {
    /**
     * Bit flags that give the timing point extra effects. See [the effects section](#effects).
     */
    kiai: boolean

    /**
     * Whether or not the first barline is omitted in osu!taiko and osu!mania
     */
    omitFirstBarLine: boolean
  }
} & ({
  type: 'uninherited'
  /**
   * The BPM for this timing point.
   */
  BPM: number
} | {
  type: 'inherited'
  /**
   * The slider velocity multiplier for this timing point.
   * For example, a `sliderMultiplier` of `2` means `BPM * 2` in osu! editor.
   */
  sliderMultiplier: number
})

export enum SampleSet {
  BeatmapDefault = 0,
  Normal = 1,
  Soft = 2,
  Drum = 3,
}
