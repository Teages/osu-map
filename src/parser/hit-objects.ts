import { useStrAutomaton } from '../utils'
import { OsuMapMode } from './general'

export function parseHitObjects(lines: string[], mode: OsuMapMode): HitObject[] {
  const items: HitObject[] = []
  lines.forEach((line) => {
    const sa = useStrAutomaton(line)

    const x = Number(sa.readUntil(','))
    const y = Number(sa.readUntil(','))
    const time = Number(sa.readUntil(','))
    const type = parseType(Number(sa.readUntil(',')), mode)
    const hitsoundStr = sa.tryReadUntil(',')

    if (type === 'circle') {
      const hitSampleStr = sa.readAll() || '0:0:0:0:'
      items.push({
        x,
        y,
        time,
        type,
        combo: parseCombo(Number(hitsoundStr)),
        hitSound: parseHitSound(hitsoundStr, hitSampleStr),
      })
    }

    else if (type === 'slider') {
      const curveStr = sa.readUntil(',')
      const [curveType, ...rawCurvePoints] = curveStr.split('|').map(v => v.trim())
      const curvePoints = rawCurvePoints.map((point) => {
        const [x, y] = point.split(':').map(v => v.trim()).map(v => Number(v))
        return { x, y }
      })
      const slides = Number(sa.readUntil(','))
      const length = Number(sa.tryReadUntil(','))

      if (sa.preview() === '') {
        // no hitsound for this note, use default
        const repeatArray = Array.from({ length: slides + 1 }).fill(0)
        sa.add(`${repeatArray.join('|')},${repeatArray.map(() => '0:0').join('|')}`)
      }

      const edgeHitSoundEnable = sa.readUntil(',').split('|')
        .map(v => v.trim())
        .map(v => Number(v))
        .map(parseHitSoundEnable)
      const edgeHitSound = sa.tryReadUntil(',').split('|').map((hitSound, index) => {
        const [normalSet, additionSet] = hitSound.split(':').map(v => v.trim())
        const sample = edgeHitSoundEnable[index]
        return {
          normalSet: parseHitSoundSet(Number(normalSet)),
          additionSet: parseHitSoundSet(Number(additionSet)),
          sample,
        }
      })
      const hitSampleStr = sa.readAll() || '0:0:0:0:'

      items.push({
        x,
        y,
        time,
        type,
        combo: parseCombo(Number(hitsoundStr)),
        hitSound: parseHitSound(hitsoundStr, hitSampleStr),

        // slider
        curveType: curveType as HitObjectSlider['curveType'],
        curvePoints,
        slides,
        length,
        edgeHitSound,
      })
    }

    else if (type === 'spinner') {
      const endTime = Number(sa.tryReadUntil(','))
      const hitSampleStr = sa.readAll() || '0:0:0:0:'
      items.push({
        x,
        y,
        time,
        type,
        combo: parseCombo(Number(hitsoundStr)),
        hitSound: parseHitSound(hitsoundStr, hitSampleStr),

        // spinner
        endTime,
      })
    }

    else { // type === 'hold'
      const endTime = Number(sa.tryReadUntil(':')) // idk why it's `:` instead of `,`

      const hitSampleStr = sa.readAll() || '0:0:0:0:'
      items.push({
        x,
        y,
        time,
        type,
        combo: parseCombo(Number(hitsoundStr)),
        hitSound: parseHitSound(hitsoundStr, hitSampleStr),

        // hold
        endTime,
      })
    }
  })

  return items
}

function parseHitSound(hitsoundStr: string, hitSampleStr: string) {
  const { normalSample, additionSimple, index, volume, filename } = parseHitSample(hitSampleStr)
  return {
    normalSet: parseHitSoundSet(normalSample),
    additionSet: parseHitSoundSet(additionSimple),
    sample: parseHitSoundEnable(Number(hitsoundStr)),
    index,
    volume,
    filename,
  }
}

function parseHitSample(hitSampleStr: string) {
  const [normalSetStr, additionSetStr, indexStr, volumeStr, filename] = hitSampleStr.split(':')
  const [normalSample, additionSimple, index, volume]
    = [normalSetStr, additionSetStr, indexStr, volumeStr].map(Number)

  return { normalSample, additionSimple, index, volume, filename: filename || undefined }
}

function parseHitSoundSet(bit: number): HitSoundSampleSet {
  switch (bit) {
    case 0:
      return 'auto'
    case 1:
      return 'normal'
    case 2:
      return 'soft'
    case 3:
      return 'drum'
    default:
      throw new Error(`Unknown sample set: ${bit}`)
  }
}

function parseHitSoundEnable(bit: number): HitObjectHitSound['sample'] {
  return {
    normal: (bit & 1) === 1,
    whistle: (bit & 2) === 2,
    finish: (bit & 4) === 4,
    clap: (bit & 8) === 8,
  }
}

function parseType(typeBit: number, mode: OsuMapMode): HitObjectType {
  if (typeBit & 0b1) {
    return 'circle'
  }
  if (typeBit & 0b10) {
    return 'slider'
  }
  if (typeBit & 0b1000) {
    return 'spinner'
  }
  if (typeBit & 0b10000000 && mode === OsuMapMode.Mania) {
    return 'hold'
  }
  throw new Error(`Unknown hit object type: ${typeBit}(mode:${mode})`)
}

function parseCombo(typeBit: number): HitObjectNewCombo {
  return {
    newCombo: (typeBit & 0b100) === 4,
    skip: (typeBit & 0b1110000) >> 4,
  }
}

export type HitObject = HitObjectCircle | HitObjectSlider | HitObjectSpinner | HitObjectHold

export type HitObjectCircle = HitObjectBasic & {
  type: 'circle'
}

export type HitObjectSlider = HitObjectBasic & {
  type: 'slider'

  /**
   * Type of curve used to construct this slider.
   *
   * `B` = b√©zier, `C` = centripetal catmull-rom, `L` = linear, `P` = perfect circle
   */
  curveType: 'B' | 'C' | 'L' | 'P'

  /**
   * Anchor points used to construct the slider.
   */
  curvePoints: Array<{
    x: number
    y: number
  }>

  /**
   * Amount of times the player has to follow the slider's curve back-and-forth before the slider is complete. It can also be interpreted as the repeat count plus one.
   */
  slides: number

  /**
   * Visual length in [osu! pixels](https://osu.ppy.sh/wiki/en/Client/Beatmap_editor/osu%21_pixel) of the slider.
   */
  length: number

  /**
   * Hitsounds that play when hitting edges of the slider's curve. The first sound is the one that plays when the slider is first clicked, and the last sound is the one that plays when the slider's end is hit.
   */
  edgeHitSound: Array<HitObjectHitSound>
}

export type HitObjectSpinner = HitObjectBasic & {
  type: 'spinner'
  endTime: number
}

export type HitObjectHold = HitObjectBasic & {
  type: 'hold'
  endTime: number
}

export interface HitObjectBasic {
  /**
   * x position
   */
  x: number

  /**
   * y position
   */
  y: number

  /**
   * Time when the object is to be hit, in milliseconds from the beginning of the beatmap's audio.
   */
  time: number

  /**
   * The combo state of the object.
   */
  combo: HitObjectNewCombo

  /**
   * The hit sound when the object is hit.
   */
  hitSound: HitObjectHitSound & {
    /**
     * Index of the sample. If this is `0`, the timing point's sample index will be used instead.
     */
    index: number

    /**
     * Volume of the sample from 1 to 100. If this is `0`, the timing point's volume will be used instead.
     */
    volume: number

    /**
     * Custom filename of the addition sound.
     */
    filename?: string
  }
}

export interface HitObjectHitSound {
  /**
   * Sample set of the normal sound.
   */
  normalSet: HitSoundSampleSet

  /**
   * Sample set of the whistle, finish, and clap sounds.
   */
  additionSet: HitSoundSampleSet

  /**
   * The sounds will play when the object is hit.
   */
  sample: {
    /**
     * It is enabled even it is `false` except in osu!mania.
     */
    normal: boolean
    whistle: boolean
    finish: boolean
    clap: boolean
  }
}

export type HitObjectNewCombo = {
  /**
   * The object is a new combo.
   */
  newCombo: true

  /**
   * Specifying how many combo colours to skip, a practice referred to as "colour hax". Only relevant if the object starts a new combo.
   */
  skip: number
} | {
  newCombo: false
}

export type HitSoundSampleSet = 'auto' | 'normal' | 'soft' | 'drum'

export type HitObjectType = 'circle' | 'slider' | 'spinner' | 'hold'
