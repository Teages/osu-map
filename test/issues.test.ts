import { describe, expect, it } from 'vitest'
import { parseHitObjects } from '../src/parser/hit-objects'
import { OsuMapMode } from '../src/parser/general'

describe('issues', () => {
  it('#1', () => {
    expect(parseHitObjects([
      '131,169,9759,2,0,P|155:177|176:192,1,52.5',
    ], OsuMapMode.Standard)).matchSnapshot()
  })
})
