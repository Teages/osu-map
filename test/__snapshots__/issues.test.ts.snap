// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`issues > #1 1`] = `
[
  {
    "combo": {
      "newCombo": false,
      "skip": 0,
    },
    "curvePoints": [
      {
        "x": 155,
        "y": 177,
      },
      {
        "x": 176,
        "y": 192,
      },
    ],
    "curveType": "P",
    "edgeHitSound": [
      {
        "additionSet": "auto",
        "normalSet": "auto",
        "sample": {
          "clap": false,
          "finish": false,
          "normal": false,
          "whistle": false,
        },
      },
      {
        "additionSet": "auto",
        "normalSet": "auto",
        "sample": {
          "clap": false,
          "finish": false,
          "normal": false,
          "whistle": false,
        },
      },
    ],
    "hitSound": {
      "additionSet": "auto",
      "filename": undefined,
      "index": 0,
      "normalSet": "auto",
      "sample": {
        "clap": false,
        "finish": false,
        "normal": false,
        "whistle": false,
      },
      "volume": 0,
    },
    "length": 52.5,
    "slides": 1,
    "time": 9759,
    "type": "slider",
    "x": 131,
    "y": 169,
  },
]
`;
